types/validator_set_test.go

```diff
@@ -20,12 +20,6 @@ func TestValidatorSetBasic(t *testing.T) {
 	// empty or nil validator lists are allowed,
 	// but attempting to IncrementProposerPriority on them will panic.
 	vset := NewValidatorSet([]*Validator{})
```

Comment:
The comment on top of this line still states that panic will still happen.

Reply:
fixed

Reply:
As discussed today I also added a check for null validator set and error after updateValidators() is called in updateState().

Reply:
And this will panic / shut down further up in `finalizeCommit` / poolRoutine. 
Not sure but the comment in the latter also sounds a bit concerning: https://github.com/tendermint/tendermint/blob/4d7b29cd8f8ef998cf8f7141f3a100b0a4fa718d/blockchain/reactor.go#L338-L342

Reply:
I've seen that also. Here it looks like we decided that recovery from all errors of ApplyBlock() is not possible. In general I think it's fine to panic in places where recovery is not possible. For validator updates related failures ( updates are invalid such that they either cannot be processed or could result in a bad state) maybe we could proceed by discarding the validator updates...not sure though.
Also, in the future we could completely validate the updates in validateValidatorUpdates(). Currently this doesn't catch all problems, some come later via updateValidators().

Reply:
The only reason we allow the validator set to be empty is during chain initialization where there are no validators in the genesis file because the application will return the initial validator set when we call `InitChain`. 

However, any use of the ValidatorSet before calling InitChain should probably panic. So I think Anton is right that we don't want to just return from these functions if they're used on an empty validator set - though we can explicitly panic instead of waiting for it to happen from a nil dereference.